pygmt/clib/conversion.py

```diff
@@ -162,19 +162,36 @@ def _to_numpy(data: Any) -> np.ndarray:
         "date64[ms][pyarrow]": np.datetime64,
     }
 
-    if (
-        hasattr(data, "isna")
-        and data.isna().any()
-        and Version(pd.__version__) < Version("2.2")
-    ):
-        # Workaround for dealing with pd.NA with pandas < 2.2.
-        # Bug report at: https://github.com/GenericMappingTools/pygmt/issues/2844
-        # Following SPEC0, pandas 2.1 will be dropped in 2025 Q3, so it's likely
-        # we can remove the workaround in PyGMT v0.17.0.
-        array = np.ascontiguousarray(data.astype(float))
-    else:
-        vec_dtype = str(getattr(data, "dtype", ""))
-        array = np.ascontiguousarray(data, dtype=dtypes.get(vec_dtype))
+    # pandas nullable dtypes and pyarrow types were converted to np.object_ dtype
+    # before, and are converted to suitable numpy dtypes since pandas 2.2.
+    # Refer to the following link for details:
+    # https://pandas.pydata.org/docs/whatsnew/v2.2.0.html#to-numpy-for-numpy-nullable-and-arrow-types-converts-to-suitable-numpy-dtype
+    # Here are the workarounds for pandas < 2.2.
+    # Following SPEC 0, pandas 2.1 should be dropped in 2025 Q3, so it's likely we can
+    # remove the workaround in PyGMT v0.17.0.
+    if Version(pd.__version__) < Version("2.2"):
+        dtypes.update(
+            {
+                "Int8": np.int8,
+                "Int16": np.int16,
+                "Int32": np.int32,
+                "Int64": np.int64,
+                "UInt8": np.uint8,
+                "UInt16": np.uint16,
+                "UInt32": np.uint32,
+                "UInt64": np.uint64,
+                "Float32": np.float32,
+                "Float64": np.float64,
+            }
+        )
+        if hasattr(data, "isna") and data.isna().any():
+            # Integer dtypes with missing values are cast to NumPy float dtypes and NaN
+            # is used as missing value indicator.
+            dtype = np.float64 if data.dtype.kind in "iu" else data.dtype.numpy_dtype
+            data = data.to_numpy(na_value=np.nan).astype(dtype=dtype)
```

Comment:
When dealing with missing values with pandas 2.1, we have some choices to convert pandas array into numpy array.

1. `np.ascontinuous(data, dtype=np.float64)`
2. `data.to_numpy(dtype=dtype, na_value=np.nan)`
3. `data.to_numpy(na_value=np.nan, dtype=dtype)`

As shown below, it turns out only the 3rd way works for all cases.

```python
In [1]: import pandas as pd
x =
In [2]: x = pd.Series([1, 2, pd.NA], dtype=pd.Int32Dtype())

In [3]: import numpy as np

In [4]: np.ascontiguousarray(x, dtype=np.float64)
...
ValueError: cannot convert to 'float64'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
In [5]: x.to_numpy(dtype=np.float64, na_value=np.nan)
Out[5]: array([ 1.,  2., nan])

In [6]: x.to_numpy(na_value=np.nan).astype(np.float64)
Out[6]: array([ 1.,  2., nan])

In [7]: x = pd.Series([1, 2, pd.NA], dtype="float16[pyarrow]")

In [8]: x
Out[8]:
0     1.0
1     2.0
2    <NA>
dtype: halffloat[pyarrow]

In [9]: x.to_numpy(dtype=np.float64, na_value=np.nan)
...
ArrowTypeError: Expected np.float16 instance

In [10]: x.to_numpy(na_value=np.nan).astype(np.float64)
Out[10]: array([ 1.,  2., nan])
```